### PR TITLE
Split `--tunnel` flag to allow setting a tunneling protocol for a subset of traffic

### PR DESCRIPTION
**Pull request by pchaigno**
This pull request splits the `--tunnel` flag into two new flags to allow setting the tunneling protocol independently of the default routing mode (tunneling or native). This is useful for e.g. the egress gateway where users may want to use GENEVE instead of VXLAN but can't today.

Summary:
- Commit 1 introduces the two new flags.
- Commits 2--6 replace all uses of the `--tunnel` flag with the new flags.
- Commit 7 deprecates the `--tunnel` flag.
- Commit 8 updates the upgrade guide.

Test runs for the three issue_comment workflows modified in this PR:
- AWS-CNI: https://github.com/cilium/cilium/actions/runs/4543365761?pr=24561
- Datapath: https://github.com/cilium/cilium/actions/runs/4543365722?pr=24561
- Clustermesh: https://github.com/cilium/cilium/actions/runs/4543365755?pr=24561

```release-note
Deprecate `--tunnel` in favor of `--routing-mode` and `--tunnel-protocol`.
```

### CODE REVIEW

File: Documentation/operations/upgrade.rst
Comment (christarazi) [inline Documentation/operations/upgrade.rst]: Did you mean v1.15?

File: pkg/option/config.go
Comment (christarazi) [inline pkg/option/config.go:3084]: Nit: Why fatal here when most* of the function is returning errors?

*: The other inconsistency is on L3310 and L3103. All others return errors.

**Review by christarazi — APPROVED** [review]
Thanks for the very easy to read PR with nicely structured commits and a clear summary; a model example! :sparkles: 

A few minor non-blocking comments.

File: Documentation/operations/upgrade.rst
Comment (julianwiedmann) [inline Documentation/operations/upgrade.rst]:
```suggestion
* Following the deprecation of the ``tunnel`` agent flag, ``tunnel`` has been
  deprecated in favor of ``routingMode`` and ``tunnelProtocol``, and will be
  removed in v1.14.
```

File: pkg/datapath/loader/base.go
Comment (julianwiedmann) [inline pkg/datapath/loader/base.go]: The comment is no longer accurate (the protocol can now be either `vxlan` or `geneve`).

File: pkg/option/config.go
Comment (julianwiedmann) [inline pkg/option/config.go]: I hope we can make this `s/TunnelingEnabled/RoutingModeIsTunnel` some day.

**Review by julianwiedmann — APPROVED** [review]
Thanks! Just a few small comments from browsing through. Could easily be addressed in a follow-on.

Re https://github.com/cilium/cilium/pull/24561/commits/55dc5914b39ef2f7ddf0b158b8f5d46c461151a5

> Options --routing-mode and --tunnel-protocol default to native and
vxlan respectively.

If I'm not mistaken, the `--routing-mode` actually defaults to `tunnel`?

**Comment by julianwiedmann** [comment]
In context of https://github.com/cilium/cilium/issues/19786, a nice follow-on now would be to switch the datapath to use `s/TUNNEL_MODE/ROUTING_MODE_TUNNEL`.

File: Documentation/operations/upgrade.rst
Comment (tklauser) [inline Documentation/operations/upgrade.rst]: Should also mention `routing-mode=tunnel` here?
```suggestion
  ``tunnel=disabled``). To configure the tunneling protocol, you can set
  ``routing-mode=tunnel`` and ``tunnel-protocol=geneve`` (previously ``tunnel=geneve``).
```

File: Documentation/operations/upgrade.rst
Comment (pchaigno) [reply Documentation/operations/upgrade.rst]: Ouch, yes. That's because this is a rebase of a PR I already attempted to send into v1.13.

**Review by tklauser — APPROVED** [review]
Nice work on the clean commit history! Only one nit in addition to what Chris and Julian already pointed out.

Also, the commit message in the 3rd commit `helm: New flags for native routing and tunnel protocol` mentions a helm flag `nativeRouting.enabled` and a `enable-native-routing` flag. I guess these should say `routingMode` and `routing-mode` instead, respectively.

File: Documentation/operations/upgrade.rst
Comment (pchaigno) [reply Documentation/operations/upgrade.rst]: `routing-mode=tunnel` is already the default value (I said something different in a commit description; it was one of Julian's comments).

File: Documentation/operations/upgrade.rst
Comment (zacharysarah) [inline Documentation/operations/upgrade.rst]: I think it would help to specify the version in which the option was deprecated. For example:

```suggestion
* The ``tunnel`` option was deprecated in v1.1<x> and will be removed in v1.14. To enable
```

File: Documentation/operations/upgrade.rst
Comment (zacharysarah) [inline Documentation/operations/upgrade.rst]: Great example of "to accomplish X, do Y" writing. A small tweak will tighten it up.
```suggestion
  ``tunnel=disabled``). To configure the tunneling protocol, set
```

File: Documentation/operations/upgrade.rst
Comment (zacharysarah) [inline Documentation/operations/upgrade.rst]: Again, I think it would help to specify the version in which the flag was deprecated. One typo fix:
```suggestion
* Following the deprecation of the ``tunnel`` agent flag in v1.1<x>, ``tunnel`` has been
```

File: Documentation/operations/upgrade.rst
Comment (zacharysarah) [reply Documentation/operations/upgrade.rst]: +1 on the typo fix, but no comma needed.

File: pkg/option/config.go
Comment (zacharysarah) [inline pkg/option/config.go]:
```suggestion
	// (tunneling is enabled) in that case.
```

File: pkg/option/config.go
Comment (zacharysarah) [reply pkg/option/config.go]: Agreed.

**Review by zacharysarah — CHANGES_REQUESTED** [review]
Nice work. ✨ Some suggestions and fixes, otherwise LGTM. 👍🏻

File: Documentation/operations/upgrade.rst
Comment (pchaigno) [reply Documentation/operations/upgrade.rst]: I'm not sure I understand this one, but maybe the confusion comes from the typo here: this sentence should say "will be removed in v1.15", as per Chris's comment above.

Then, this whole section of the guide is about v1.14. Do we really want to specify that each item in the list is happening in v1.14?

Comment (pchaigno) [reply Documentation/operations/upgrade.rst]: Hm, maybe the tense is the issue here. I keep using past tense when the other entries are using present or present continuous. Maybe that would help clarify we're still talking about v1.14?

Comment (zacharysarah) [reply Documentation/operations/upgrade.rst]: > maybe the confusion comes from the typo here: this sentence should say "will be removed in v1.15", as per Chris's comment above.

There's no confusion about the version in which the option will be _removed_. I saw Chris's feedback before I reviewed. My suggestion was specifically to indicate the version in which the option was _deprecated_ (presumably v1.14).

> Then, this whole section of the guide is about v1.14. Do we really want to specify that each item in the list is happening in v1.14?

If the entire context of the document is explicitly specific to v1.14 and unlikely to be reused in a different, non-specific context, then I think it's fine to ignore my suggestion.

However: looking at the file, I see nothing to indicate a specific release in the title or description. Rather, I see that the page is implied to be version-specific because of the branch in which the file will reside when merged. 

The challenge, then, is that when this page is updated in 1.15 and subsequent versions, there will be nothing to indicate the version in which the options were removed--because the version in which deprecation occurred is implied, not explicit.

Does that make sense?

> I keep using past tense when the other entries are using present or present continuous. 

It's not a problem of verb tense agreement; it's a problem of implied vs. explicit. 

> Maybe that would help clarify we're still talking about v1.14?

I certainly think so. 😄

**Review by ldelossa — APPROVED** [review]
Looks good to me, and a welcome clarification of the flags.

I still see one reference to to `tunnel=disabled` in 

```
test/bigtcp/test.sh
34:    --set tunnel='disabled' \
```

Worth checking out?

File: Documentation/operations/upgrade.rst
Comment (pchaigno) [reply Documentation/operations/upgrade.rst]: This document isn't all specific to v1.14, but the section where those changes are in is for v1.14 only. See line 312 where the title is (this is a lot clearer on the rendered docs than on GitHub of course).

Once we branch for v1.15, a new v1.15 section will be added here. These entries will not be reused for other versions. This PR is also on purpose not marked for backport so it won't appear in older stable branches.

I'm not against being more specific on versions in sentences if we still think it's needed, but that doesn't look specific to this change. I'd prefer if we address it for all entries in a separate PR if we go that direction.

I'll fix up the bad verb tense.

File: pkg/option/config.go
Comment (pchaigno) [reply pkg/option/config.go:3084]: @christarazi Hm, I'm missing something here. There are 19 other fatals in this same function. The function also doesn't return anything anyway.

File: Documentation/operations/upgrade.rst
Comment (joestringer) [reply Documentation/operations/upgrade.rst]: (FWIW I agree with @zacharysarah here that we should just be more explicit, but yeah seems fine to just update them all at once in another PR to set that expectation moving forwards)

**Review by zacharysarah — APPROVED** [review]

File: pkg/option/config.go
Comment (tommyp1ckles) [inline pkg/option/config.go]: minor: Should we include valid modes in the error message

**Review by tommyp1ckles — APPROVED** [review]
lg 👏

**Review by nbusseneau — APPROVED** [review]
Changes LGTM, are you also taking care of making changes accordingly to the Cilium CLI?

**Comment by julianwiedmann** [comment]
Think this needs a rebase for https://github.com/cilium/cilium/pull/23890 (specifically the changes in `test/k8s`).

**Comment by julianwiedmann** [comment]
(FWIW, the `ipsec` codeowner was never assigned for this PR and is still dangling :/)

**Comment by pchaigno** [comment]
> Changes LGTM, are you also taking care of making changes accordingly to the Cilium CLI?

@nbusseneau Yes, I will, once this is merged.

File: pkg/option/config.go
Comment (christarazi) [reply pkg/option/config.go:3084]: Oh I think I was looking at the wrong function, nevermind.

**Comment by pchaigno** [comment]
/test-1.26-net-next

**Comment by pchaigno** [comment]
/test-1.24-5.4

**Comment by pchaigno** [comment]
/test-1.24-5.4

**Review by nebril — APPROVED** [review]

**Comment by pchaigno** [comment]
/test-1.26-net-next

**Comment by pchaigno** [comment]
/test-1.26-net-next

**Comment by pchaigno** [comment]
/test-1.26-net-next

**Comment by julianwiedmann** [comment]
Re https://github.com/cilium/cilium/pull/24561/commits/522ece3cfbe297d6679a2c1d3da23727733e82b6 - I take it we hit this issue when using `--tunnel-port`? Nice catch.

I was holding off on https://github.com/cilium/cilium/pull/24647 due to slight concerns about the restart becoming more disruptive (we're ripping out the old encap interface, but haven't updated `ENCAP_IFINDEX` in the BPF programs yet). But not worth fretting over when it fixes an actual bug :).

**Comment by macmiranda** [comment]
> > Changes LGTM, are you also taking care of making changes accordingly to the Cilium CLI?
> 
> @nbusseneau Yes, I will, once this is merged.

Did anyone ever get to this? I'm still getting pre-1.14 helm values with the latest CLI version.

**Pull request merged**
